go/tools/builders/asm.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2017 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+type params struct {
+	// Path to the "go" binary in the Go toolchain.
+	GoTool string `json:"go_tool"`
+
+	// List of directories that may contain headers included by the source.
+	Includes []string `json:"includes"`
+
+	// The Go assembly source file to build.
+	Source string `json:"source"`
+
+	// Path to the .o file that the assembler should write.
+	Out string `json:"out"`
+}
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Fprintf(os.Stderr, "usage: %s params\n", os.Args[0])
+		os.Exit(1)
+	}
+
+	paramsData := []byte(os.Args[1])
+	var p params
+	if err := json.Unmarshal(paramsData, &p); err != nil {
+		log.Fatalf("error parsing params file: %v", err)
```

Comment:
optional: for unit-testing, I have always preffered a style of a short main which catches the error and log.Fatalf, and a
func run(argv) error method that returns error in all of the places where you have log.Fatalf here.